From 551ef1562dd8854522a598bf80ca5725f6391043 Mon Sep 17 00:00:00 2001
From: Marc Mutz <marc.mutz@qt.io>
Date: Thu, 16 Oct 2025 18:00:46 +0200
Subject: [PATCH 442/553] =?UTF-8?q?QChar:=20avoid=20char{32=E2=86=9216}=5F?=
 =?UTF-8?q?t=20narrowing=20in=20canonicalOrderHelper()?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Same pattern as in the previous few patches: the old code used a
char32_t to temporarily store a UTF-16 code point, narrowing it back
to char16_t in a call to QChar::surrogateToUcs4(), triggering Clang 21
-Wcharacter-conversion.

Solution is also the same: keep the UTF-16 code unit in a separate
variable, until we have determined whether it needs surrogate decoding
or merely promotion to char32_t.

Amends the start of the public history.

Pick-to: 6.8 6.5
Change-Id: I955948348d637c4fe13485808bef47a4ee58f7bf
Reviewed-by: Ahmad Samir <a.samirh78@gmail.com>
(cherry picked from commit 72663b81177aa17b3ebfeaff63a687729194f6d7)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/corelib/text/qchar.cpp | 20 ++++++++++++++------
 1 file changed, 14 insertions(+), 6 deletions(-)

diff --git a/src/corelib/text/qchar.cpp b/src/corelib/text/qchar.cpp
index 167ba9e84d0..6be19473b85 100644
--- a/src/corelib/text/qchar.cpp
+++ b/src/corelib/text/qchar.cpp
@@ -1999,26 +1999,34 @@ static void canonicalOrderHelper(QString *str, QChar::UnicodeVersion version, qs
     qsizetype pos = from;
     while (pos < l) {
         qsizetype p2 = pos+1;
-        char32_t u1 = s.at(pos).unicode();
-        if (QChar::isHighSurrogate(u1)) {
+        char32_t u1;
+        if (const char16_t hi = s.at(pos).unicode(); QChar::isHighSurrogate(hi)) {
             const char16_t low = s.at(p2).unicode();
             if (QChar::isLowSurrogate(low)) {
-                u1 = QChar::surrogateToUcs4(u1, low);
+                u1 = QChar::surrogateToUcs4(hi, low);
                 if (p2 >= l)
                     break;
                 ++p2;
+            } else {
+                u1 = hi;
             }
+        } else {
+            u1 = hi;
         }
         ushort c1 = 0;
 
     advance:
-        char32_t u2 = s.at(p2).unicode();
-        if (QChar::isHighSurrogate(u2) && p2 < l) {
+        char32_t u2;
+        if (const char16_t hi = s.at(p2).unicode(); QChar::isHighSurrogate(hi) && p2 < l) {
             const char16_t low = s.at(p2+1).unicode();
             if (QChar::isLowSurrogate(low)) {
-                u2 = QChar::surrogateToUcs4(u2, low);
+                u2 = QChar::surrogateToUcs4(hi, low);
                 ++p2;
+            } else {
+                u2 = hi;
             }
+        } else {
+            u2 = hi;
         }
 
         ushort c2 = 0;
-- 
2.51.2

